# Homework 1

Robert Rose

In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv("./Parking_Citations.csv")

/home/rob/.pyenv/versions/anaconda3-5.3.0/lib/python3.7/site-packages/IPython/core/interactiveshell.py:2785: DtypeWarning: Columns (2) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


The [Open Baltimore website](https://data.baltimorecity.gov/Transportation/Parking-Citations/n4ma-fj3m) says that this dataset contains information on parking citations issued within the City. Data represents years of citations and any older citations with outstanding balances. It is updated every weekday automatically by the Finance Department.

In [3]:
df.head(50)

Citation      Tag ExpMM  ExpYY State   Make  \
0   82691651  69943CF    04   19.0    MD   MAZD   
1   85937226  T720887    12   18.0    MD   ACUR   
2   82691115  5BA2678    03   19.0    MD   TOYT   
3   82694689  5BT8544    01   18.0    MD   FORD   
4   82691214  1BV0537    11   18.0    MD   SUBA   
5   82688467  5CZ9237    08   19.0    MD   NISS   
6   82690141   MCG966    12   19.0    MD   HOND   
7   82688442  9DE2544    03   20.0    MD   NISS   
8   82693525  7DA8010    10   19.0    MD   HYUN   
9   82687725  4DA3646    08   19.0    MD   JEEP   
10  82689150   6DCX96    10   19.0    MD   TOYT   
11  82692832  5AP5656    02   20.0    MD   TOYT   
12  82697393  89508CF    07   18.0    MD   TOYT   
13  85937242  5DB4425    09   19.0    MD   CHEV   
14  85937259  38270CH    03   18.0    MD   HOND   
15  82689739  7DE9147    03   20.0    MD   NISS   
16  82685372  2AN3165    07   18.0    MD   CHEV   
17  82692360   BRT002    05   18.0    MI  TOYOT   
18      1537  7AD3147    01   13.0    MD   CHRY   
19  82696254   KHY369    03   19.0    MD   HOND   
20  82694135  28758M6    02   19.0    MD    KIA   
21  82694655  3DE2314    02   20.0    MD   MAZD   
22  82696510  097M578    07   19.0    MD   HOND   
23  82692410   PP9H4L    08   18.0    MO   HYUN   
24  82688392  BKT4450    04   19.0    NC    BMW   
25  82688582  55622CG    06   19.0    MD   HYUN   
26  82692584  43688CG    10   18.0    MD   LEXS   
27  82692147  8CW5981    04   19.0    MD   DODG   
28  82696536  6CM0625    09   18.0    MD   DODG   
29  82691073  50956CD    05   18.0    MD   HOND   
30  85937275  2BE2371    12   19.0    MD   NISS   
31      2485   244084    03   12.0    DE    BMW   
32      2527   8FHL65    05   12.0    MD  ACURA   
33      2568  5AJ4941    08   13.0    MD   BUIC   
34      2600  3AM1741    01   12.0    MD   FORD   
35  82694119  2CZ4377    07   19.0    MD   HOND   
36  82689804  2CT3000    12   18.0    MD   HOND   
37  82690026  UB00069    01   19.0    MD    RAM   
38  82693038  3BT0398    07   18.0    MD    KIA   
39  85937291  2CV2541    03   19.0    MD   HYUN   
40  82692519  8DB0804    11   19.0    MD   TOYT   
41  82687923   2EST11    12   19.0    MD    BMW   
42  82691198  2CD6446    11   18.0    MD   MERZ   
43  82696346  5BM8056    06   18.0    MD   MAZD   
44  82690307  7DA1615    08   19.0    MD   VOLK   
45  82690588  1CG7321    03   18.0    MD   HYUN   
46  82691990  9CH2190    06   18.0    MD   HYUN   
47  82687493  GYG2936    05   17.0    NY  LEXUS   
48  82687576  ZGL0029    07   19.0    MD   TOYT   
49  82691560  6DE2226    01   20.0    MD   MERZ   

                           Address  ViolCode  \
0   1300 BLK EAST NORTHERN PKWY WB        32   
1         300 BLK NORTH BEND RD SB        32   
2           5000 BLK ROLAND AVE NB        32   
3           2400 BLK ERDMAN AVE NB        32   
4   1300 BLK WEST COLD SPRING LN W        32   
5       3400 BLKLOCH RAVEN BLVD NB        32   
6       6100 BLK OLD HARFORD RD NB        32   
7           4300 BLK ERDMAN AVE NB        32   
8           5600 BLK MORAVIA RD NB        32   
9           4300 BLK ERDMAN AVE NB        32   
10    2700 BLK GWYNNS FALLS PKY WB        32   
11     1200 BLK SOUTH CATON AVE NB        32   
12          2400 BLK ERDMAN AVE NB        32   
13     2800 BLK REISTERSTOWN RD WB        32   
14     2800 BLK REISTERSTOWN RD EB        32   
15          2400 BLK ERDMAN AVE NB        32   
16     1200 BLK SOUTH CATON AVE NB        32   
17          2400 BLK ERDMAN AVE NB        32   
18                      UB S EUTAW        12   
19          4401 PENNINGTON AVE SB        32   
20          2400 BLK ERDMAN AVE NB        32   
21  1300 BLK WEST COLD SPRING LN E        32   
22       4400 BLK EDMONDSON AVE WB        32   
23          5000 BLK ROLAND AVE NB        32   
24         5900 BLK WALTHER AVE NB        32   
25  1300 BLK EAST NORTHERN PKWY WB        32   
26          4300 BLK ERDMAN AVE NB        32   
27     1200 BLK SOUTH CATON AVE NB        32   
28  1300 B

In [4]:
df.tail(50)

Citation      Tag ExpMM  ExpYY State   Make  \
2908561  90523912  03457LM     3   15.0    MD  LINCO   
2908562  81435043  2CP1933    11   18.0    MD   ACUR   
2908563  82278475  1AC6339     9   18.0    MD   HYUN   
2908564  81637036  39978CG    12   18.0    MD   TOYT   
2908565  62748034  6BX2331     4   19.0    MD  CHEVR   
2908566  60554897  2AT5414     6   18.0    MD   FORD   
2908567  80107700  5CR2641    12   18.0    MD   TOYT   
2908568  80393623  5AY7014    11   18.0    MD   SUBA   
2908569  81153281  9CS7370    12   18.0    MD   HYUN   
2908570  82659419  9AL9665    12   17.0    MD   CHEV   
2908571  80937965  3BV7471     2   19.0    MD   CHEV   
2908572  62448726  7AV3666    10   18.0    MD   FORD   
2908573  80288260  4CS4907    12   18.0    MD   NISS   
2908574  80779987  4CN3007    10   18.0    MD   JEEP   
2908575  80758858  6CX0211     4   19.0    MD   MERZ   
2908576  62113965  E333856    11   17.0    IL  CHEVR   
2908577   4619863  4CT8314     2   19.0    MD   FORD   
2908578  62313672  25890CD     2   19.0    MD  HONDA   
2908579  80309868  3AX8135    10   18.0    MD   NISS   
2908580  81806037  9AT2594    11   18.0    MD   NISS   
2908581  60901007  5AY7014    11   18.0    MD  SUBAR   
2908582  60251338     3326     3   18.0    MD  HONDA   
2908583  61625274  3AV0891     6   16.0    MD  LINCO   
2908584  80407092  6CR1583    12   18.0    MD   NISS   
2908585  80193916  3CP5518    10   18.0    MD   DODG   
2908586  62580288  1DE4921     1   20.0    MD  NISSA   
2908587  60642916  1AY4706    11   18.0    MD  CHEVR   
2908588  62405676  06146AE     3   15.0    MD    BMW   
2908589  63168901  9BA0475     5   19.0    MD  CHEVR   
2908590  80991715  5CR2641    12   18.0    MD   TOYT   
2908591  81821077  9AT2594    11   18.0    MD   NISS   
2908592  61186285  C0MPETE     5   19.0    MD  ACURA   
2908593  80161251  85602CG     1   19.0    MD   NISS   
2908594  61879152  9CS0113    12   18.0    MD    KIA   
2908595  62787271  JNK0757     4   17.0    PA  NISSA   
2908596  61902244  4CB3056    10   17.0    MD  HONDA   
2908597   4879541  7DC8057    11   19.0    MD   TOYT   
2908598  62290557  K33655K   NaN    0.0    PA  CHEVR   
2908599  63118195  E333856    11   17.0    IL  CHEVR   
2908600  82001497  8CS5658    12   18.0    MD   HOND   
2908601  22257398  M00RISH     3   14.0    MD  CHEVR   
2908602  60516219  JGC9828    10   17.0    TX  TOYOT   
2908603  62638359  T200721   NaN    0.0    TN  TRAIL   
2908604  80479414  5CR5419    12   18.0    MD   CHRY   
2908605  61713377  T200721   NaN    0.0    TN  TRAIL   
2908606  62352050  8CR4708    10   17.0    MD   FORD   
2908607  80095442   1843Z8     9   18.0    MD   DODG   
2908608  90464885  N0SIDAM     3   16.0    MD  NISSA   
2908609  82614067   MTH893    12   17.0    MD   HYUN   
2908610  80694228  1CT5429    11   18.0    MD   TOYT   

                                Address  ViolCode  \
2908561                    600 PRATT ST        99   
2908562           1200 BLK E 33RD ST WB        32   
2908563           1200 BLK E 33RD ST WB        32   
2908564          4300 BLK ERDMAN AVE NB        32   
2908565                 1100 CARROLL ST        12   
2908566             200 UNIVERSITY PKWY        12   
2908567       4400 BLK EDMONDSON AVE EB        32   
2908568     1200 BLK SOUTH CATON AVE NB        32   
2908569       4400 BLK EDMONDSON AVE WB        32   
2908570     1200 BLK SOUTH CATON AVE NB        32   
2908571           1200 BLK E 33RD ST WB        32   
2908572                    900 LIGHT ST        18   
2908573       4400 BLK EDMONDSON AVE EB        32   
2908574       4400 BLK EDMONDSON AVE EB        32   
2908575     1200 BLK SOUTH CATON AVE NB        32   
2908576                600 LEXINGTON ST        18   
2908577                   800 W 36TH ST        18   
2908578                   500 EXETER ST        26   
2908579    2700 BLK GWYNNS FALLS PKY WB        32   
2908580       2700 BLK WATERVIEW AVE WB        32   
2908581                 1600 ST PAUL ST

## Head/Tail Quality Comments

From just the `head()` and `tail()` commands, it looks like the data is pretty good for several of the columns (`Citation`, `Tag`, `ExpMM`, `ExpYY`, `State`, `Make`, `Address`, `ViolCode`, `Description`, `ViolFine`, `Balance`, `OpenFine`, `OpenPenalty`, `ImportDate` and `Location`) but is missing pretty much universally for several of the other columns (`Neighborhood`, `PoliceDistrict`, `CouncilDistrict`). It looks like some of the columns are only conditionally filled like `PenaltyDate` and `NoticeDate`).

## Column Content

I chose to work with the columns that were totally filled in. Those Columns ended up being `State`, `ViolCode`, `ViolFine`, `OpenFine`, `OpenPenalty` and `Balance`. Although `ImportDate` and `Description` were also not missing any values, I chose not do anything with them because it looked like `Description` was just the prose description of the violation type, and therefore not very interesting to play with and `ImportDate` didn't seem to mean much about the data, more so when the data was entered.

In [5]:
df.isnull().sum()

Citation                 0
Tag                    298
ExpMM                72785
ExpYY                   18
State                    0
Make                  2841
Address                  2
ViolCode                 0
Description              0
ViolFine                 0
ViolDate              4530
Balance                  0
PenaltyDate        2908611
OpenFine                 0
OpenPenalty              0
NoticeDate         2113713
ImportDate               0
Neighborhood       2583092
PoliceDistrict     2583092
CouncilDistrict    2583067
Location            743502
dtype: int64

### State

#### Description

This is the origin State for the license plate of the violation. There are no missing datapoints.

#### Outliers

Since this is a categorical value, I don't think there will be any outliers persay. I did want to get the number of unique values in the dataframe however. You can see that bit below.

In [16]:
unique_states = df.groupby('State')['State'].count()
display(unique_states.sort_values(ascending=False))

State
MD    2499766
VA      76138
PA      75864
NJ      27762
NY      23665
NC      20114
FL      17899
DE      16907
DC      16883
TN      15171
TX      13095
GA      10273
CA       7620
OH       7609
SC       7306
MA       6790
WV       6734
ME       6683
CT       6009
IL       5373
MI       5259
AZ       3430
IN       3180
AL       2735
MO       1859
OK       1833
WA       1726
LA       1723
CO       1609
OR       1528
       ...   
AE          1
YU          1
GE          1
HS          1
WS          1
WE          1
VR          1
VM          1
VJ          1
VI          1
VE          1
UN          1
UK          1
TS          1
TL          1
TA          1
ST          1
SE          1
QU          1
PI          1
NW          1
MV          1
YT          1
MP          1
MC          1
LS          1
KX          1
KI          1
KA          1
00          1
Name: State, Length: 125, dtype: int64

#### Observations

Unsurprisingly, the most common state of vehicle to get tickets in Baltimore is Maryland. Next most common are Virginia and Pennsylvania, two states that border Maryland. I was surprised how many 1-ticket states there were. I was assuming there would be around 50 but there's actually 125! I assume a lot of those are Reservations which issue their own license plates.

### ViolCode

#### Description

This is the violation number that was written on the ticket. Each violation corresponds to a description. For example, `32` is for `Fixed Speed Camera`'s.

#### Outliers

There is no missing data, I made sure to only use columns with no missing data. I also don't really think there will be any outliers since this is a categorical variable, but I decided to look at the number of each violation code regardless.

In [17]:
unique_violcodes = df.groupby('ViolCode')['ViolCode'].count()
display(unique_violcodes.sort_values(ascending=False))

ViolCode
32    734392
18    553575
27    256946
30    238133
22    209699
12    189824
8     171194
11    126237
31     82275
99     75350
33     56631
5      33457
16     30140
17     20364
15     19413
20     18769
19     18063
1      15257
26     13056
3      12728
46      8371
10      6391
34      4921
6       4394
43      2722
13      1863
9       1261
42       758
47       689
41       521
25       458
23       321
28       196
4         98
2         86
45        26
50        10
40         9
44         8
52         2
14         1
7          1
51         1
Name: ViolCode, dtype: int64

#### Observations

As expected there's no missing data, but I am curious as to what the violcodes with very few tickets correspond to.

In [21]:
df['Description'].loc[(df['ViolCode'] == 51) | (df['ViolCode'] == 7) | (df['ViolCode'] == 14) | (df['ViolCode'] == 52)]

2419       No Stopping/Parking Stadium Event – 33rd
196189                 Res. Park Permit 4th Offense
1602756                Res. Park Permit 3rd Offense
1810697      Parking or Parking Meter Tow-Away zone
1810701                Res. Park Permit 4th Offense
Name: Description, dtype: object

Alright, it looks like they correspond to a few different tickets that have happened only a few times in the dataset. I do wonder though how there can be two "4th Offense" tickets and only one "3rd Offense".